### PR TITLE
Kafka integration

### LINKED ISSUE #936: Feat: Add native Kafka integration for Parseable server
User should be able to set environment variables to integrate their Kafka cluster with Parseable. 

- Decide which env vars are needed for this integration.
- Use a Rust Kafka client like https://github.com/kafka-rust/kafka-rust to subscribe to events from the Kafka client.
- Ingest data from the given Kafka topic.
- Once events are ingested, merge the binary data from Kafka to the Parseable data ingestion flow.

### PR DESCRIPTION
**Pull request by TomBebb**
/claim #936

Fixes #936

### Description

Integrates with Kakfa, using its crate.
This is toggled on/off via the KAFKA_TOPIC  environment variable.
This uses a background polling iterator to loop through messages, and ingesting messages to the stream with the name of the KAFKA_TOPIC.
Covers all possible config options via environment variables.

<hr>

This PR has:
- [ y] been tested to ensure log ingestion and log query works.
- [ y] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ y] added documentation for new or modified features or behaviors.

### CODE REVIEW

**Comment by TomBebb** [comment]
Kafka-console-producer:
Screenshot_20241108_120807

Data:
Screenshot_20241108_120822

File: server/src/kafka.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        parsed_timestamp: Utc::now().naive_utc(),
+        time_partition: None,
+        custom_partition_values: HashMap::new(),
+        stream_type: StreamType::UserDefined,
+    }
+    .process_unchecked()?;
+    Ok(())
+}
+
+pub async fn setup_integration() -> Result<JoinHandle<()>, KafkaError> {
```
Comment (de-sh) [inline server/src/kafka.rs:193]: It should be fine to drop the handle if I am not wrong, why are we returning it?

File: server/Cargo.toml [hunk 2]
```diff
@@ -106,6 +106,8 @@ path-clean = "1.0.1"
 prost = "0.13.3"
 prometheus-parse = "0.2.5"
 sha2 = "0.10.8"
+kafka = "0.10.0"
```
Comment (de-sh) [inline server/Cargo.toml:109]: Using https://github.com/fede1024/rust-rdkafka instead of kafka-rust could allow the implementation to be async

**Comment by de-sh** [comment]
@TomBebb can you please document the code bit more, this will help us understand what we are going to maintain in the long run 😅 

The code here is lacking comments, especially regarding the functions and what they are doing as well as the use of good const names that describes what they are doing. Thank you in advance!

File: server/src/kafka.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+        let res = task::spawn(async move {
+            let consumer = setup_consumer().unwrap();
+            let mut stream = consumer.stream();
+            loop {
+                while let Some(curr) = stream.next().await {
+                    let msg = curr.unwrap();
+                    ingest_message(&stream_name, msg).await.unwrap();
+                }
+            }
```
Comment (de-sh) [inline server/src/kafka.rs:223]: `while let` inside a `loop`? Also, why the unwrap is not after `stream.next()` and instead on the returned value when [ref](https://github.com/fede1024/rust-rdkafka/blob/master/src/consumer/stream_consumer.rs#L317) says to do otherwise?

File: .gitignore [hunk 4]
```diff
@@ -14,4 +14,5 @@ parseable
 parseable_*
 parseable-env-secret
 cache
+.idea
```
Comment (de-sh) [inline .gitignore:17]: please remove unrelated change

File: .idea/.gitignore
Comment (de-sh) [inline .idea/.gitignore:1]: same as above

File: server/src/kafka.rs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        }
+        .process()
+        .await?;
+    } else {
+        log::debug!("{} No payload for stream", stream_name);
+    }
+    Ok(())
+}
+
+pub async fn setup_integration() -> Result<JoinHandle<()>, KafkaError> {
```
Comment (de-sh) [inline server/src/kafka.rs:210]:
```suggestion
pub async fn setup_integration() -> Result<(), KafkaError> {
```

File: server/src/kafka.rs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    let my_res = if let Ok(stream_name) = env::var("KAFKA_TOPIC") {
+        log::info!("Setup kafka integration for {stream_name}");
+        create_stream_if_not_exists(&stream_name, &StreamType::UserDefined.to_string()).await?;
+
+        let res = task::spawn(async move {
```
Comment (de-sh) [inline server/src/kafka.rs:215]:
```suggestion
        task::spawn(async move {
```

File: server/src/kafka.rs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        .process()
+        .await?;
+    } else {
+        log::debug!("{} No payload for stream", stream_name);
+    }
+    Ok(())
+}
+
+pub async fn setup_integration() -> Result<JoinHandle<()>, KafkaError> {
+    let my_res = if let Ok(stream_name) = env::var("KAFKA_TOPIC") {
```
Comment (de-sh) [inline server/src/kafka.rs:211]:
```suggestion
    if let Ok(stream_name) = env::var("KAFKA_TOPIC") {
```

File: server/src/kafka.rs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                    let msg = curr.unwrap();
+                    ingest_message(&stream_name, msg).await.unwrap();
+                }
+            }
+        });
+        log::info!("Done Setup kafka integration");
+        res
+    } else {
+        task::spawn_blocking(|| {})
+    };
```
Comment (de-sh) [inline server/src/kafka.rs:229]:
```suggestion
    }
```

File: server/src/kafka.rs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    Ok(my_res)
```
Comment (de-sh) [inline server/src/kafka.rs:230]:
```suggestion
    Ok(())
```

**Comment by parmesant** [comment]
Hey @TomBebb !
I wanted to try out your changes but am running into some issues

**Steps**
- setup kafka from binary and start the server
- set `KAFKA_TOPIC` env var
- set `KAFKA_HOSTS` env var

image

- run parseable (local-store) - I can see the logstream get created on the console
image

- start a bootstrap server and send a JSON message - I don't see any debug logs on parseable or any log events on console

image

- close the bootstrap server - I still don't see any logs on console
image


Did I miss some step?
